src/appleseed/renderer/modeling/postprocessingstage/bloompostprocessingstage.cpp

```diff
@@ -0,0 +1,425 @@
+
+//
+// This source file is part of appleseed.
+// Visit https://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2020 Tiago Chaves, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "bloompostprocessingstage.h"
+
+// appleseed.renderer headers.
+#include "renderer/global/globallogger.h"
+#include "renderer/modeling/frame/frame.h"
+#include "renderer/modeling/postprocessingstage/effect/additiveblendapplier.h"
+#include "renderer/modeling/postprocessingstage/effect/brightpassapplier.h"
+#include "renderer/modeling/postprocessingstage/effect/resampleapplier.h"
+#include "renderer/modeling/postprocessingstage/effect/resamplex2applier.h"
+#include "renderer/modeling/postprocessingstage/postprocessingstage.h"
+
+// appleseed.foundation headers.
+#include "foundation/containers/dictionary.h"
+#include "foundation/image/canvasproperties.h"
+#include "foundation/image/color.h"
+#include "foundation/image/image.h"
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/types.h"
+#include "foundation/utility/api/apistring.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+
+// Standard headers.
+#include <cstddef>
+#include <cmath>
+#include <memory>
+#include <vector>
+
+using namespace foundation;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // Bloom post-processing stage.
```

Comment:
You should explain the general algorithm somewhere (passes process and stuff). Be sure to also link your references